File: services/device/src/main/java/com/example/iter/device/support/EquipmentImageUrlResolver.java
package com.example.iter.device.support;

import com.example.iter.common.storage.S3StorageProperties;
import com.example.iter.device.domain.entity.EquipmentImage;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;
import software.amazon.awssdk.services.s3.S3Client;
import software.amazon.awssdk.services.s3.model.GetUrlRequest;

@Component
@RequiredArgsConstructor
public class EquipmentImageUrlResolver {

    private final S3Client s3Client;
    private final S3StorageProperties properties;

    public String resolve(EquipmentImage image) {
        if (!StringUtils.hasText(image.getObjectKey())) {
            return image.getImageUrl();
        }
        return resolve(image.getObjectKey());
    }

    public String resolve(String objectKey) {
        if (StringUtils.hasText(properties.getPublicBaseUrl())) {
            return stripTrailingSlash(properties.getPublicBaseUrl()) + "/" + objectKey;
        }
        return s3Client.utilities()
                .getUrl(GetUrlRequest.builder()
                        .bucket(properties.getBucket())
                        .key(objectKey)
                        .build())
                .toExternalForm();
    }

    private String stripTrailingSlash(String value) {
        String result = value;
        while (result.endsWith("/")) {
            result = result.substring(0, result.length() - 1);
        }
        return result;
    }
}

File: services/device/src/main/kotlin/com/example/iter/device/support/EquipmentImageUrlResolver.kt
package com.example.iter.device.support

import com.example.iter.common.storage.S3StorageProperties
import com.example.iter.device.domain.entity.EquipmentImage
import org.springframework.stereotype.Component
import org.springframework.util.StringUtils
import software.amazon.awssdk.services.s3.S3Client
import software.amazon.awssdk.services.s3.model.GetUrlRequest

@Component
class EquipmentImageUrlResolver(
    private val s3Client: S3Client,
    private val properties: S3StorageProperties,
) {

    fun resolve(image: EquipmentImage): String {
        if (!StringUtils.hasText(image.objectKey)) {
            return image.imageUrl
        }
        return resolve(image.objectKey!!)
    }

    fun resolve(objectKey: String): String {
        if (StringUtils.hasText(properties.publicBaseUrl)) {
            return properties.publicBaseUrl!!.trimEnd('/') + "/" + objectKey
        }
        return s3Client.utilities()
            .getUrl(
                GetUrlRequest.builder()
                    .bucket(properties.bucket)
                    .key(objectKey)
                    .build()
            )
            .toExternalForm()
    }
}
